src/cls/rgw_gc/cls_rgw_gc.cc

```diff
@@ -276,60 +310,50 @@ static int cls_rgw_gc_queue_remove_entries(cls_method_context_t hctx, bufferlist
     unsigned int index = 0;
     // If data is not empty
     if (op_ret.entries.size()) {
-      for (auto it : op_ret.entries) {
+      for (const auto& it : op_ret.entries) {
         cls_rgw_gc_obj_info info;
         try {
           decode(info, it.data);
         } catch (ceph::buffer::error& err) {
           CLS_LOG(5, "ERROR: cls_rgw_gc_queue_remove_entries(): failed to decode gc info\n");
           return -EINVAL;
         }
+        if (!last_removed_time || *last_removed_time < info.time) {
+          last_removed_time = info.time;
+        }
         CLS_LOG(20, "INFO: cls_rgw_gc_queue_remove_entries(): entry: %s\n", info.tag.c_str());
         total_num_entries++;
         index++;
-        bool found = false;
         //Search for tag in urgent data map
-        auto iter = urgent_data.urgent_data_map.find(info.tag);
-        if (iter != urgent_data.urgent_data_map.end()) {
-          found = true;
-          if (iter->second > info.time) {
-            CLS_LOG(10, "INFO: cls_rgw_gc_queue_remove_entries(): tag found in urgent data: %s\n", info.tag.c_str());
+        if (auto i = urgent_data.urgent_data_map.find(info.tag);
+            i != urgent_data.urgent_data_map.end()) {
+          const auto defer_time = i->second;
+          CLS_LOG(10, "INFO: cls_rgw_gc_queue_remove_entries(): erasing tag from urgent data: %s\n", info.tag.c_str());
+          urgent_data.urgent_data_map.erase(i);
+
+          if (defer_time > info.time) {
```

Comment:
defer_time, can be lesser than the expiration time of the last entry in the queue, since this time is based on when defer gc was called. I think it will be better to calculate this time based on now + expiration_seconds.

Reply:
`expiration_secs` is from the client's configuration, and isn't part of the `cls_rgw_gc_queue_remove_entries_op` so we don't have access to it here

thinking about the case where the queue is otherwise empty: if we requeue it with `defer_time`, then it's eligible for GC then. requeueing it with `now + expiration_secs` would add an extra delay before it can be removed

in the non-empty case: we may requeue a `defer_time` that's smaller than other entries in the queue, so it won't be eligible for GC until those other entries expire. but it's also smaller than entries enqueued after, so won't ever delay processing of later entries

can you see any issue with the expiration timestamp going back in time between entries? i feel like we have to deal with cases like that, because radosgws can change their configuration for `expiration_secs`

Reply:
If the queues can have mixed timestamp, then we will have to make sure that in list_entries we break out as soon as we encounter a non expired entry when we are listing expired entries only. If there are 100 entries in queue and the 50th one has a lesser timestamp, and all are non expired except the 50th one, then going by current code, we will skip all non expired ones and pick the expired one and send that back to the user. Since the remove call takes only num entries as input, it will remove the first entry erroneously.